### PR TITLE
mempool: length prefix txs when getting them from mempool

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

In protobuf `[]byte` is varint encoded. When adding txs to the block we were not taking this into account. 

Closes: #XXX

### CODE REVIEW

File: mempool/clist_mempool.go [hunk 1]
```diff
@@ -527,11 +528,13 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
 	txs := make([]types.Tx, 0, mem.txs.Len())
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
 		memTx := e.Value.(*mempoolTx)
+		txLength, _ := binary.Varint(memTx.tx)
+		txSize := txLength + int64(len(memTx.tx))
 		// Check total size requirement
-		if maxBytes > -1 && totalBytes+int64(len(memTx.tx)) > maxBytes {
+		if maxBytes > -1 && totalBytes+int64(len(memTx.tx))+txSize > maxBytes {
```
Comment (melekes) [inline mempool/clist_mempool.go:534]:
```suggestion
		if maxBytes > -1 && totalBytes+txSize > maxBytes {
```

File: mempool/clist_mempool.go [hunk 2]
```diff
@@ -527,11 +528,13 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
...
 			return txs
 		}
-		totalBytes += int64(len(memTx.tx))
+		totalBytes += int64(len(memTx.tx)) + txSize
```
Comment (melekes) [inline mempool/clist_mempool.go:537]:
```suggestion
		totalBytes += txSize
```

File: node/node_test.go [hunk 3]
```diff
@@ -305,10 +305,13 @@ func TestCreateProposalBlock(t *testing.T) {
 	}
 	assert.EqualValues(t, partSetFromHeader.ByteSize(), partSet.ByteSize())
 
+<<<<<<< HEAD
```
Comment (melekes) [inline node/node_test.go]: ??

File: node/node_test.go [hunk 4]
```diff
@@ -272,7 +272,7 @@ func TestCreateProposalBlock(t *testing.T) {
 	// fill the mempool with more txs
 	// than can fit in a block
 	txLength := 100
-	for i := 0; i <= maxBytes/txLength; i++ {
+	for i := 0; i < maxBytes/txLength; i++ {
```
Comment (marbar3778) [inline node/node_test.go:275]:
```suggestion
	for i := 0; i <= maxBytes/txLength; i++ {
```

File: types/tx.go [hunk 5]
```diff
@@ -137,3 +138,14 @@ func TxProofFromProto(pb tmproto.TxProof) (TxProof, error) {
 
 	return pbtp, nil
 }
+
+func ComputeProtoOverhead(tx Tx, fieldNum int) int64 {
```
Comment (marbar3778) [inline types/tx.go:142]: add godoc

File: types/tx.go [hunk 6]
```diff
@@ -137,3 +138,18 @@ func TxProofFromProto(pb tmproto.TxProof) (TxProof, error) {
 
 	return pbtp, nil
 }
+
+// ComputeProtoOverhead calculates the overhead for protobuf encoing of a transaction.
```
Comment (cmwaters) [inline types/tx.go:142]:
```suggestion
// ComputeProtoOverhead calculates the overhead for protobuf encoding of a transaction.
```

File: mempool/clist_mempool.go [hunk 7]
```diff
@@ -527,11 +527,14 @@ func (mem *CListMempool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
 	txs := make([]types.Tx, 0, mem.txs.Len())
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
 		memTx := e.Value.(*mempoolTx)
+
+		txLength := types.ComputeProtoOverhead(memTx.tx, 1)
```
Comment (melekes) [inline mempool/clist_mempool.go:531]:
```suggestion
		protoOverhead := types.ComputeProtoOverhead(memTx.tx, 1)
```

File: mempool/mempool.go [hunk 8]
```diff
@@ -105,7 +105,9 @@ type TxInfo struct {
 // PreCheckMaxBytes checks that the size of the transaction is smaller or equal to the expected maxBytes.
 func PreCheckMaxBytes(maxBytes int64) PreCheckFunc {
 	return func(tx types.Tx) error {
-		txSize := int64(len(tx))
+		txLength := types.ComputeProtoOverhead(tx, 1)
```
Comment (melekes) [inline mempool/mempool.go:108]:
```suggestion
		protoOverhead := types.ComputeProtoOverhead(tx, 1)
```

File: types/tx.go [hunk 9]
```diff
@@ -137,3 +138,18 @@ func TxProofFromProto(pb tmproto.TxProof) (TxProof, error) {
 
 	return pbtp, nil
 }
+
+// ComputeProtoOverhead calculates the overhead for protobuf encoding of a transaction.
+// The overhead consists of varint encoding the field number and the wire type
+// (= length-delimited = 2), and another varint encoding the length of the
+// transaction. https://developers.google.com/protocol-buffers/docs/encoding
+func ComputeProtoOverhead(tx Tx, fieldNum int) int64 {
```
Comment (melekes) [inline types/tx.go:146]:
```suggestion
func ComputeProtoOverheadForTx(tx Tx, fieldNum int) int64 {
```

File: mempool/mempool.go (hunk 8 above)
Comment (melekes) [inline mempool/mempool.go:108]: I wonder if we can just use https://github.com/golang/protobuf/blob/91c84e0db17890c2bb64280a6d660e73e4237fd1/proto/wire.go#L13 `Size` function or other func exposed by golang protobuf. Have you thought about it?

Comment (marbar3778) [reply mempool/mempool.go:108]: Yes, looked for a bit but to do this we would have to wrap the tx in proto.Message then call size on it. This way we avoid doing this

Comment (melekes) [reply mempool/mempool.go]: But if the protobuf spec changes, we won't have to change our code

Comment (erikgrinaker) [reply mempool/mempool.go]: Yeah, I'm inclined to agree with @melekes here.

Comment (marbar3778) [reply mempool/mempool.go]: made the change, running tests locally. I agree as well

**Comment by erikgrinaker** [comment]
When this ends up in the block, will it take into account the varint that gives the length of the transaction list? https://developers.google.com/protocol-buffers/docs/encoding#packed

In the simplest case, I suppose it can just assume a worst-case overhead of 10 bytes, which isn't all that much to waste.

**Comment by marbar3778** [comment]
> When this ends up in the block, will it take into account the key of each repeated tx?
> [developers.google.com/protocol-buffers/docs/encoding#optional](https://developers.google.com/protocol-buffers/docs/encoding#optional)

no it does not, I can make the needed change. In testing this has not come up because there was not greater that 5k txs in a block, good catch.

**Comment by erikgrinaker** [comment]
What does 5k signify?

**Comment by marbar3778** [comment]
> What does 5k signify?

Amount of txs in a block.

**Comment by erikgrinaker** [comment]
Yes, but why is 5k significant? Why would it work up until 5k but then break?

**Comment by marbar3778** [comment]
> Yes, but why is 5k significant? Why would it work up until 5k but then break?

Sorry, meant we didn't observe this with 5k txs in a block, (an example from testing on a running network). I didn't mean it in the sense that it would work till 5k then break.

File: types/tx.go [hunk 10]
```diff
@@ -137,3 +139,10 @@ func TxProofFromProto(pb tmproto.TxProof) (TxProof, error) {
 
 	return pbtp, nil
 }
+
+// ComputeProtoOverheadForTx calculates the size for a protobuf encoded transaction.
+// https://developers.google.com/protocol-buffers/docs/encoding
+func ComputeProtoOverheadForTx(tx Tx) int64 {
```
Comment (cmwaters) [inline types/tx.go:145]: Do you mind tweaking the function name. It no longer calculates the overhead but the entire size of the proto-encoded transaction

**Review by cmwaters — APPROVED** [review]
:+1: Good stuff

**Review by erikgrinaker — APPROVED** [review]

File: types/tx.go [hunk 11]
```diff
@@ -137,3 +137,23 @@ func TxProofFromProto(pb tmproto.TxProof) (TxProof, error) {
 
 	return pbtp, nil
 }
+
+// ComputeProtoOverheadForTxs wraps the transactions in tmproto.Data{} and calculates the size.
+// https://developers.google.com/protocol-buffers/docs/encoding
+func ComputeProtoSizeForTxs(txs []Tx, tx Tx) int64 {
```
Comment (melekes) [inline types/tx.go:143]: why have tx Tx as a second parameter when you can add it to `txs`? `ComputeProtoSizeForTxs(txs, tx))`

Comment (marbar3778) [reply types/tx.go:143]: Will open a pr for this change

**Pull request merged**